File: samples/dotnet/BgeM3.Onnx/M3EmbedderConfig.cs
using Microsoft.ML.OnnxRuntime;

namespace BgeM3.Onnx;

/// <summary>
/// Configuration for M3Embedder initialization
/// </summary>
public record M3EmbedderConfig
{
    /// <summary>
    /// Primary execution provider to use
    /// </summary>
    public ExecutionProvider ExecutionProvider { get; init; } = ExecutionProvider.CPU;

    /// <summary>
    /// Fallback execution providers in order of preference
    /// </summary>
    public ExecutionProvider[] FallbackProviders { get; init; } = [ExecutionProvider.CPU];

    /// <summary>
    /// CUDA device ID (only used when ExecutionProvider is CUDA)
    /// </summary>
    public int CudaDeviceId { get; init; } = 0;

    /// <summary>
    /// Enable memory pattern optimization
    /// </summary>
    public bool EnableMemoryPattern { get; init; } = true;

    /// <summary>
    /// Enable CPU memory arena
    /// </summary>
    public bool EnableCpuMemArena { get; init; } = true;

    /// <summary>
    /// Log severity level (0=Verbose, 1=Info, 2=Warning, 3=Error, 4=Fatal)
    /// </summary>
    public OrtLoggingLevel LogSeverityLevel { get; init; } = OrtLoggingLevel.ORT_LOGGING_LEVEL_WARNING;
}

File: samples/java/bge-m3-onnx/src/main/java/com/yunikosoftware/bgem3onnx/M3EmbedderConfig.java
package com.yunikosoftware.bgem3onnx;

/**
 * Configuration for M3Embedder initialization
 */
public class M3EmbedderConfig {
    private final ExecutionProvider executionProvider;
    private final ExecutionProvider[] fallbackProviders;
    private final int cudaDeviceId;
    private final boolean enableMemoryPattern;
    private final boolean enableCpuMemArena;
    private final int logSeverityLevel;

    public M3EmbedderConfig() {
        this(ExecutionProvider.CPU, new ExecutionProvider[]{ExecutionProvider.CPU}, 0, true, true, 2);
    }

    public M3EmbedderConfig(ExecutionProvider executionProvider, ExecutionProvider[] fallbackProviders, 
                           int cudaDeviceId, boolean enableMemoryPattern, boolean enableCpuMemArena, 
                           int logSeverityLevel) {
        this.executionProvider = executionProvider;
        this.fallbackProviders = fallbackProviders;
        this.cudaDeviceId = cudaDeviceId;
        this.enableMemoryPattern = enableMemoryPattern;
        this.enableCpuMemArena = enableCpuMemArena;
        this.logSeverityLevel = logSeverityLevel;
    }

    public ExecutionProvider getExecutionProvider() {
        return executionProvider;
    }

    public ExecutionProvider[] getFallbackProviders() {
        return fallbackProviders;
    }

    public int getCudaDeviceId() {
        return cudaDeviceId;
    }

    public boolean isEnableMemoryPattern() {
        return enableMemoryPattern;
    }

    public boolean isEnableCpuMemArena() {
        return enableCpuMemArena;
    }

    public int getLogSeverityLevel() {
        return logSeverityLevel;
    }

    public static class Builder {
        private ExecutionProvider executionProvider = ExecutionProvider.CPU;
        private ExecutionProvider[] fallbackProviders = new ExecutionProvider[]{ExecutionProvider.CPU};
        private int cudaDeviceId = 0;
        private boolean enableMemoryPattern = true;
        private boolean enableCpuMemArena = true;
        private int logSeverityLevel = 2;

        public Builder executionProvider(ExecutionProvider executionProvider) {
            this.executionProvider = executionProvider;
            return this;
        }

        public Builder fallbackProviders(ExecutionProvider... fallbackProviders) {
            this.fallbackProviders = fallbackProviders;
            return this;
        }

        public Builder cudaDeviceId(int cudaDeviceId) {
            this.cudaDeviceId = cudaDeviceId;
            return this;
        }

        public Builder enableMemoryPattern(boolean enableMemoryPattern) {
            this.enableMemoryPattern = enableMemoryPattern;
            return this;
        }

        public Builder enableCpuMemArena(boolean enableCpuMemArena) {
            this.enableCpuMemArena = enableCpuMemArena;
            return this;
        }

        public Builder logSeverityLevel(int logSeverityLevel) {
            this.logSeverityLevel = logSeverityLevel;
            return this;
        }

        public M3EmbedderConfig build() {
            return new M3EmbedderConfig(executionProvider, fallbackProviders, cudaDeviceId, 
                                      enableMemoryPattern, enableCpuMemArena, logSeverityLevel);
        }
    }
}